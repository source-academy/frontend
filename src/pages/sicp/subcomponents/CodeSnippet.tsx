import { Card, Elevation, Pre } from '@blueprintjs/core';
import { HighlightRulesSelector, ModeSelector } from 'js-slang/dist/editors/ace/modes/source';
import { Resizable } from 're-resizable';
import * as React from 'react';
import { PrismLight as SyntaxHighlighter } from 'react-syntax-highlighter';
import ControlBar from 'src/commons/controlBar/ControlBar';
import { ControlBarCloseButton } from 'src/commons/controlBar/ControlBarCloseButton';
import { useResponsive } from 'src/commons/utils/Hooks';
import { SourceTheme } from 'src/features/sicp/SourceTheme';

import { CodeSnippetContext } from '../Sicp';
import SicpWorkspaceContainer from './SicpWorkspaceContainer';

export type CodeSnippetProps = OwnProps;
type OwnProps = {
  body: string;
  output: string;
  id: string;
  initialEditorValueHash: string;
  prependLength: number | undefined;
};

const resizableProps = {
  enable: {
    top: false,
    right: false,
    bottom: true,
    left: false,
    topRight: false,
    bottomRight: false,
    bottomLeft: false,
    topLeft: false
  },
  defaultSize: {
    width: '100%',
    height: '500px'
  },
  minHeight: '250px',
  maxHeight: '2000px'
};

const CodeSnippet: React.FC<CodeSnippetProps> = props => {
  const { body, output, id } = props;
  const context = React.useContext(CodeSnippetContext);
  const { isMobileBreakpoint } = useResponsive();

  const handleOpen = () => {
    context.setActive(id);
  };

  const handleClose = React.useCallback(() => {
    context.setActive('0');
  }, [context]);

  const WorkspaceProps = {
    initialEditorValueHash: props.initialEditorValueHash,
    prependLength: props.prependLength,
    isSicpEditor: true,

    handleCloseEditor: handleClose
  };

  HighlightRulesSelector(4);
  ModeSelector(4);

  const closeButton = React.useMemo(
    () => <ControlBarCloseButton key="close" handleClose={handleClose} />,
    [handleClose]
  );

  const controlBarProps = {
    editorButtons: [],
    flowButtons: [],
    editingWorkspaceButtons: [closeButton]
  };

  return (
    <div className="sicp-code-snippet">
      {context.active === id ? (
        <div className="sicp-code-snippet-open">
          <ControlBar {...controlBarProps} />
          {isMobileBreakpoint ? (
            <div className="sicp-workspace-container-container">
              <SicpWorkspaceContainer {...WorkspaceProps} />
            </div>
          ) : (
            <div className="sicp-code-snippet-desktop-open">
              <Resizable {...resizableProps}>
                <div className="sicp-workspace-container-container">
                  <SicpWorkspaceContainer {...WorkspaceProps} />
                </div>
              </Resizable>
            </div>
          )}
        </div>
      ) : (
        <Card className="sicp-code-snippet-closed" interactive={true} elevation={Elevation.TWO}>
          <SyntaxHighlighter language="javascript" style={SourceTheme} onClick={handleOpen}>
            {body}
          </SyntaxHighlighter>
        </Card>
      )}
      {output && (
        <Pre>
          <em>{output}</em>
        </Pre>
      )}
    </div>
  );
};

export default CodeSnippet;
