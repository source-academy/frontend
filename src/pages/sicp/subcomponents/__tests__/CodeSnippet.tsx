import 'ace-builds';
import 'ace-builds/webpack-resolver';
import 'ace-builds/src-noconflict/ace';

import { shallow } from 'enzyme';
import lzString from 'lz-string';

import CodeSnippet from '../CodeSnippet';

describe('Sicp Code Snippet', () => {
  const body = 'const a = 1;\na+1;';
  const output = '2';
  const program = lzString.compressToEncodedURIComponent(body);

  test('renders correctly with prepend', () => {
    const props = {
      body: body,
      output: output,
      id: 'id',
      initialEditorValueHash: program,
      prependLength: 1
    };

    const tree = shallow(<CodeSnippet {...props} />);
    expect(tree.debug()).toMatchSnapshot();
  });

  test('renders correctly without prepend', () => {
    const props = {
      body: body,
      output: output,
      id: 'id',
      initialEditorValueHash: program,
      prependLength: 0
    };

    const tree = shallow(<CodeSnippet {...props} />);
    expect(tree.debug()).toMatchSnapshot();
  });
});
