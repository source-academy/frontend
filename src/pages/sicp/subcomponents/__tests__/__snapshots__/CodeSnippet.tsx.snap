// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sicp Code Snippet renders correctly with prepend 1`] = `
"<div className=\\"sicp-code-snippet\\">
  <Blueprint4.Card className=\\"sicp-code-snippet-closed\\" interactive={true} elevation={2}>
    <SyntaxHighlighter language=\\"javascript\\" style={{...}} onClick={[Function: handleOpen]}>
      const a = 1;
      a+1;
    </SyntaxHighlighter>
  </Blueprint4.Card>
  <Component>
    <em>
      2
    </em>
  </Component>
</div>"
`;

exports[`Sicp Code Snippet renders correctly without prepend 1`] = `
"<div className=\\"sicp-code-snippet\\">
  <Blueprint4.Card className=\\"sicp-code-snippet-closed\\" interactive={true} elevation={2}>
    <SyntaxHighlighter language=\\"javascript\\" style={{...}} onClick={[Function: handleOpen]}>
      const a = 1;
      a+1;
    </SyntaxHighlighter>
  </Blueprint4.Card>
  <Component>
    <em>
      2
    </em>
  </Component>
</div>"
`;
